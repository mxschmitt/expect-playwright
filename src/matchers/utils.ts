import type { Page, ElementHandle, Frame } from "playwright-core"
import { PageWaitForSelectorOptions } from "../../global"

export type ExpectInputType = Page | Frame | ElementHandle

const isElementHandle = (value: ExpectInputType): value is ElementHandle => {
  return value.constructor.name === "ElementHandle"
}

export const getFrame = async (value: ExpectInputType) => {
  return isElementHandle(value) ? value.contentFrame() : value
}

const isObject = (value: unknown) =>
  typeof value === "object" && !(value instanceof RegExp)

export type InputArguments = [
  Page | ElementHandle,
  string?,
  (string | PageWaitForSelectorOptions)?,
  PageWaitForSelectorOptions?
]

export const getElementHandle = async (
  args: InputArguments,
  valueArgCount = 1
) => {
  // Pluck the options off the end first
  const options =
    args.length > 1 && isObject(args[args.length - 1])
      ? (args.pop() as PageWaitForSelectorOptions)
      : {}

  // Next, pluck the number of args required by the matcher (defaults to 1)
  const expectedValue = args.splice(-valueArgCount, valueArgCount) as string[]

  // Finally, we can find the element handle
  let elementHandle = (await getFrame(args[0])) ?? args[0]

  // If the user provided a page or iframe, we need to locate the provided
  // selector or the `body` element if none was provided.
  if (!isElementHandle(elementHandle)) {
    const selector = args[1] ?? "body"

    try {
      elementHandle = (await elementHandle.waitForSelector(selector, options))!
    } catch (err) {
      throw new Error(`Timeout exceed for element ${quote(selector)}`)
    }
  }

  return [elementHandle, expectedValue] as const
}

export const quote = (val: string | null) => (val === null ? "" : `'${val}'`)

export const getMessage = (
  { isNot, promise, utils, expand }: jest.MatcherContext,
  matcher: string,
  expected: unknown,
  received: unknown,
  expectedHint: string | undefined = undefined
) => {
  const message = isNot
    ? `Expected: not ${utils.printExpected(expected)}`
    : utils.printDiffOrStringify(
        expected,
        received,
        "Expected",
        "Received",
        expand
      )

  return (
    utils.matcherHint(matcher, undefined, expectedHint, { isNot, promise }) +
    "\n\n" +
    message
  )
}

export const compareText = (
  expectedValue: string | RegExp,
  actualValue: string | null
) => {
  return typeof expectedValue === "string"
    ? expectedValue === actualValue
    : expectedValue.test(actualValue ?? "")
}
