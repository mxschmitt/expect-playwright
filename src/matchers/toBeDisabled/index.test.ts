import { assertSnapshot } from "../tests/utils"

describe("toBeDisabled", () => {
  afterEach(async () => {
    await page.setContent("")
  })

  it("positive", async () => {
    await page.setContent('<button id="foo" disabled>')
    await expect(page).toBeDisabled("#foo")
  })

  it("negative: target element isn't enabled", async () => {
    await page.setContent('<button id="foo">')
    await assertSnapshot(() => expect(page).toBeDisabled("#foo"))
  })

  it("negative: target element not found", async () => {
    await page.setContent('<button id="foo">')
    await assertSnapshot(() =>
      expect(page).toBeDisabled("#bar", { timeout: 1000 })
    )
  })

  describe("with 'not' usage", () => {
    it("positive", async () => {
      await page.setContent('<button id="foo">')
      await expect(page).not.toBeDisabled("#foo")
    })

    it("negative", async () => {
      await page.setContent('<button id="foo" disabled>')
      await assertSnapshot(() => expect(page).not.toBeDisabled("#foo"))
    })
  })

  describe("timeout", () => {
    it("positive: should be able to use a custom timeout", async () => {
      setTimeout(() => page.setContent('<button id="foo" disabled>'), 500)
      await expect(page).toBeDisabled("#foo", { timeout: 1000 })
    })

    it("should throw an error after the timeout exceeds", async () => {
      const start = new Date().getTime()
      await assertSnapshot(() =>
        expect(page).toBeDisabled("#foo", { timeout: 1000 })
      )
      const duration = new Date().getTime() - start
      expect(duration).toBeGreaterThan(1000)
      expect(duration).toBeLessThan(1500)
    })
  })
})
