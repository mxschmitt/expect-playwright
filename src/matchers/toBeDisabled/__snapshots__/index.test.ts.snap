// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toBeDisabled negative: target element isn't enabled 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoBeDisabled[2m()[22m

Expected: [32mtrue[39m
Received: [31mfalse[39m"
`;

exports[`toBeDisabled negative: target element not found 1`] = `"Error: Timeout exceed for element '#bar'"`;

exports[`toBeDisabled timeout should throw an error after the timeout exceeds 1`] = `"Error: Timeout exceed for element '#foo'"`;

exports[`toBeDisabled with 'not' usage negative 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mnot[2m.[22mtoBeDisabled[2m()[22m

Expected: not [32mtrue[39m"
`;
